File: po_wo_pr/po_wo_pr/page/inward_dashboard/inward_dashboard.js
frappe.pages['inward-dashboard'].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Inward Document Dashboard",
		single_column: true
	});

	new InwardDashboard(page);
};

class InwardDashboard {
	constructor(page) {
		this.page = page;

		// Project Filter
		this.project = this.page.add_field({
			label: __("Project"),
			fieldname: "project",
			fieldtype: "Link",
			options: "IRS Project",
			change: () => {
				this.load_subjects();
				this.load_subject_pie_chart();
				this.refresh();
			}
		});

		// Subject Filter
		this.subject = this.page.add_field({
			label: __("Subject"),
			fieldname: "subject",
			fieldtype: "Select",
			options: [""],
			change: () => this.refresh()
		});

		// Month Filter (dropdown of month names)
		this.month = this.page.add_field({
			label: __("Month"),
			fieldname: "month",
			fieldtype: "Select",
			options: [
				"", "January", "February", "March", "April", "May", "June",
				"July", "August", "September", "October", "November", "December"
			],
			change: () => this.refresh()
		});

		// Academic Year Filter
		this.academic_year = this.page.add_field({
			label: __("Academic Year"),
			fieldname: "academic_year",
			fieldtype: "Link",
			options: "Academic Year",
			change: () => this.refresh()
		});

		this.body = $(`
			<div class="inward-dashboard">
				<div class="cards-row" style="display:flex;gap:16px;margin-bottom:20px;"></div>

				<div class="charts-row" style="display:flex;flex-wrap:wrap;gap:24px;">

					<div class="chart-box" id="status-chart" style="width:45%;"></div>

					<div class="chart-box" id="workflow-chart" style="width:45%;"></div>

					<div class="chart-box" id="medium-chart" style="width:45%;"></div>
					<div class="chart-box" id="rejection-chart" style="width:45%;"></div>
					<div class="chart-box" id="missing-docs-chart" style="width:45%;"></div>

					<div class="chart-box" id="location-chart" style="width:45%;">
						<select id="location-level" style="margin-bottom:10px;">
							<option value="district">District</option>
							<option value="taluka">Taluka</option>
							<option value="state">State</option>
						</select>
					</div>

					<div class="chart-box" id="trend-chart" style="width:94%;"></div>

					<div class="chart-box" id="subject-pie-chart" style="width:45%;"></div>

				</div>
			</div>
		`).appendTo(this.page.body);

		this.body.find("#location-level").on("change", (e) => {
			this.load_location_chart(e.target.value);
		});

		this.refresh();
	}

	load_subjects() {
		const project = this.project.get_value();

		this.subject.df.options = [""];
		this.subject.set_value("");
		this.subject.refresh();

		if (!project) {
			return;
		}

		frappe.call({
			method: "frappe.client.get",
			args: {
				doctype: "IRS Project",
				name: project
			}
		}).then(r => {

			if (!r.message) return;

			let options = [""];

			(r.message.subjects || []).forEach(row => {
				if (row.subject) {
					options.push(row.subject);
				}
			});

			this.subject.df.options = options;
			this.subject.refresh();
		});
	}

	load_rejection_chart() {
		this.call("get_rejection_reason_summary").then(data => {
			this.render_bar("rejection-chart", "By Reason of Rejection", data);
		});
	}
	load_missing_docs_chart() {
		this.call("get_missing_documents_summary").then(data => {
			this.render_bar("missing-docs-chart", "Missing / Not Received Documents", data);
		});
	}
	refresh() {
		this.load_cards();
		this.load_status_chart();
		this.load_workflow_chart();
		this.load_subject_pie_chart();
		this.load_medium_chart();
		this.load_location_chart(this.body.find("#location-level").val() || "district");
		this.load_trend_chart();
		this.load_rejection_chart();
		this.load_missing_docs_chart();
	}

	call(method, args = {}) {
		args.project = this.project.get_value();
		args.subject = this.subject.get_value();
		args.month = this.month.get_value();
		args.academic_year = this.academic_year.get_value();

		return frappe.call({
			method: "po_wo_pr.po_wo_pr.page.inward_dashboard.inward_dashboard." + method,
			args: args
		}).then(r => r.message);
	}

	load_cards() {
		this.call("get_number_cards").then(data => {
			const cards = [
				{ label: "Total", value: data.total, color: "#5e64ff" },
				{ label: "Pending", value: data.pending, color: "#ffa00a" },
				{ label: "Accepted", value: data.approved, color: "#28a745" },
				{ label: "Rejected", value: data.rejected, color: "#e03131" },
				{ label: "Repeat Rejected", value: data.repeat_rejected, color: "#9c1f1f" }
			];

			this.body.find(".cards-row").html(
				cards.map(card => `
                <div style="flex:1;padding:16px;border-radius:8px;background:${card.color}15;border-left:4px solid ${card.color};">
                    <div style="font-size:24px;font-weight:600;">${card.value || 0}</div>
                    <div style="color:#666;">${card.label}</div>
                </div>
            `).join("")
			);
		});
	}

	load_status_chart() {
		this.call("get_status_summary").then(data => {
			this.render_bar("status-chart", "By Status", data);
		});
	}

	load_workflow_chart() {
		this.call("get_application_status_summary").then(data => {
			this.render_bar("workflow-chart", "By Application Status (Accept / Reject / Repeat Reject)", data);
		});
	}

	load_medium_chart() {
		this.call("get_medium_summary").then(data => {
			this.render_bar("medium-chart", "By Medium", data);
		});
	}

	load_location_chart(level) {
		this.call("get_location_summary", {
			level: level
		}).then(data => {
			this.render_bar(
				"location-chart",
				`By ${frappe.utils.to_title_case(level)}`,
				data
			);
		});
	}

	load_subject_pie_chart() {
		const project = this.project.get_value();

		const wrapper = this.body.find("#subject-pie-chart")[0];

		// Clear chart when no project is selected
		if (!project) {
			$(wrapper).empty();
			return;
		}

		this.call("get_subject_summary").then(data => {
			$(wrapper).empty();

			const container = $("<div class='chart-container'></div>")
				.appendTo(wrapper)[0];

			new frappe.Chart(container, {
				title: "Applications by Subject",
				data: {
					labels: data.map(d => d.label),
					datasets: [{
						values: data.map(d => d.count)
					}]
				},
				type: "pie",
				height: 300
			});
		});
	}

	load_trend_chart() {
		this.call("get_monthly_trend", {
			months: 6
		}).then(data => {

			const wrapper = this.body.find("#trend-chart")[0];
			$(wrapper).find(".chart-container").remove();

			const container = $("<div class='chart-container'></div>")
				.appendTo(wrapper)[0];

			new frappe.Chart(container, {
				title: "Monthly Trend",
				data: {
					labels: data.map(d => d.label),
					datasets: [{
						name: "Documents",
						values: data.map(d => d.count)
					}]
				},
				type: "line",
				height: 220,
				colors: ["#5e64ff"]
			});
		});
	}

	render_bar(box_id, title, data) {
		const wrapper = this.body.find(`#${box_id}`)[0];

		$(wrapper).find(".chart-container").remove();

		const container = $("<div class='chart-container'></div>")
			.appendTo(wrapper)[0];

		new frappe.Chart(container, {
			title: title,
			data: {
				labels: data.map(d => d.label || "Unknown"),
				datasets: [{
					name: "Count",
					values: data.map(d => d.count)
				}]
			},
			type: "bar",
			height: 220,
			colors: ["#5e64ff"]
		});
	}
}

File: po_wo_pr/po_wo_pr/page/inward_dashboard/inward_dashboard.py
import frappe
from frappe.utils import getdate


@frappe.whitelist()
def get_number_cards(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    return frappe.db.sql(f"""
        SELECT
            COUNT(*) AS total,
            SUM(CASE WHEN application_status='Pending' THEN 1 ELSE 0 END) AS pending,
            SUM(CASE WHEN application_status='Accept' THEN 1 ELSE 0 END) AS approved,
            SUM(CASE WHEN application_status='Reject' THEN 1 ELSE 0 END) AS rejected,
            SUM(CASE WHEN application_status='Repeat Reject' THEN 1 ELSE 0 END) AS repeat_rejected
        FROM `tabInward Document`
        {where}
    """, values, as_dict=True)[0]


@frappe.whitelist()
def get_status_summary(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    return frappe.db.sql(f"""
        SELECT application_status AS label,
               COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY application_status
        ORDER BY count DESC
    """, values, as_dict=True)


@frappe.whitelist()
def get_application_status_summary(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    status_condition = "application_status IN ('Accept', 'Reject', 'Repeat Reject')"

    if where:
        where += f" AND {status_condition}"
    else:
        where = f"WHERE {status_condition}"

    return frappe.db.sql(f"""
        SELECT application_status AS label,
               COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY application_status
        ORDER BY count DESC
    """, values, as_dict=True)


@frappe.whitelist()
def get_medium_summary(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    if where:
        where += " AND medium IS NOT NULL"
    else:
        where = "WHERE medium IS NOT NULL"

    return frappe.db.sql(f"""
        SELECT medium AS label,
               COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY medium
        ORDER BY count DESC
    """, values, as_dict=True)


@frappe.whitelist()
def get_location_summary(level="district", project=None, subject=None, month=None, academic_year=None):

    if level not in ("district", "taluka", "state"):
        frappe.throw("Invalid level")

    where, values = get_conditions(project, subject, month, academic_year)

    if where:
        where += f" AND `{level}` IS NOT NULL"
    else:
        where = f"WHERE `{level}` IS NOT NULL"

    return frappe.db.sql(f"""
        SELECT `{level}` AS label,
               COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY `{level}`
        ORDER BY count DESC
        LIMIT 20
    """, values, as_dict=True)


@frappe.whitelist()
def get_monthly_trend(months=6, project=None, subject=None, month=None, academic_year=None):

    months = int(months)

    where, values = get_conditions(project, subject, month, academic_year)

    values["months"] = months

    if where:
        where += " AND received_date >= DATE_SUB(CURDATE(), INTERVAL %(months)s MONTH)"
    else:
        where = "WHERE received_date >= DATE_SUB(CURDATE(), INTERVAL %(months)s MONTH)"

    return frappe.db.sql(f"""
        SELECT
            DATE_FORMAT(received_date, '%%Y-%%m') AS label,
            COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY label
        ORDER BY label
    """, values, as_dict=True)


MONTH_NAME_TO_NUMBER = {
    "January": 1, "February": 2, "March": 3, "April": 4,
    "May": 5, "June": 6, "July": 7, "August": 8,
    "September": 9, "October": 10, "November": 11, "December": 12
}


def get_conditions(project=None, subject=None, month=None, academic_year=None):
    conditions = []
    values = {}

    if project:
        conditions.append("project = %(project)s")
        values["project"] = project

    if subject:
        conditions.append("subject = %(subject)s")
        values["subject"] = subject

    if month:
        month_number = MONTH_NAME_TO_NUMBER.get(month)
        if month_number:
            conditions.append("MONTH(received_date) = %(month)s")
            values["month"] = month_number

    if academic_year:
        ay = frappe.db.get_value(
            "Academic Year", academic_year, ["start_date", "end_date"], as_dict=True
        )
        if ay and ay.start_date and ay.end_date:
            conditions.append("received_date BETWEEN %(ay_start)s AND %(ay_end)s")
            values["ay_start"] = ay.start_date
            values["ay_end"] = ay.end_date

    where = ""
    if conditions:
        where = "WHERE " + " AND ".join(conditions)

    return where, values

@frappe.whitelist()
def get_rejection_reason_summary(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    reject_condition = "application_status IN ('Reject', 'Repeat Reject')"

    if where:
        where += f" AND {reject_condition} AND reason_for_rejection IS NOT NULL"
    else:
        where = f"WHERE {reject_condition} AND reason_for_rejection IS NOT NULL"

    return frappe.db.sql(f"""
        SELECT reason_for_rejection AS label,
               COUNT(*) AS count
        FROM `tabInward Document`
        {where}
        GROUP BY reason_for_rejection
        ORDER BY count DESC
    """, values, as_dict=True)

@frappe.whitelist()
def get_missing_documents_summary(project=None, subject=None, month=None, academic_year=None):
    where, values = get_conditions(project, subject, month, academic_year)

    if where:
        where += " AND ifnull(det.received, 0) = 0"
    else:
        where = "WHERE ifnull(det.received, 0) = 0"

    return frappe.db.sql(f"""
        SELECT det.document_name AS label,
               COUNT(*) AS count
        FROM `tabInward Details` det
        INNER JOIN `tabInward Document` id ON det.parent = id.name
        {where}
        GROUP BY det.document_name
        ORDER BY count DESC
    """, values, as_dict=True)